planet/Objects.ocd/Animals.ocd/Bat.ocd/Script.c

```diff
@@ -61,155 +61,169 @@ private func Place(int amount, proplist rectangle, proplist settings)
+            if(this.Target->GetMaterial(0, 0) != Material("Tunnel") && Time->IsDay())
+            {
+                this.Target->Fly();
+            }
+            // Start flying by random.
+            if(!Random(150))
+            {
+                this.Target->Fly();
+            }
+            return FX_OK;
+        }
+        
+        /*-- Basic actions done while flying --*/
+        
+        // Remember direction we need to fly in to get back into the home cave.
+        if(this.Target->GetMaterial(0, 0) == Material("Tunnel"))
+        {
+            this.last_cave = [this.Target->GetX(), this.Target->GetY()];
+        }
+        
+        if(!PathFree(this.Target->GetX(), this.Target->GetY(), this.last_cave[0], this.last_cave[1]) || Distance(this.Target->GetX(), this.Target->GetY(), this.last_cave[0], this.last_cave[1] > 30))
+        {
+            this.last_cave = [nil, nil];
+        }
+        
+        // Move the bat and update its in-flight appearance.
+        this.Target->ChangeMovementVelocity(this.move_direction[0], this.move_direction[1]);
+        this.Target->UpdateFlightAppearance();
+        
+        // Make flying sounds and sometimes do a sonar wave.
+        if(!Random(250))
+        {
+            this.Target->Sound("Animals::Bat::Flutter*");
+        }
+        
+        if(!Random(225))
+        {
+            this.Target->DoSonarWave(false);
+        }
+        
+        // Update the state of the bat.
+        if(this.time_since_startle > 0)
+        {
+            this.time_since_starte--;
+        }
+        
+        /*-- Random actions and actions based on events --*/
+        
+        // Only do these action updates every five frames.
+        if((time % 5) != 0) return FX_OK;
+        
+        // Stay out of water.
+        if(this.Target->GBackLiquid(0, 4) || this.Target->GBackLiquid(0, 8))
+        {
+            this.move_direction = [RandomX(-1, 1), -1];
+            return FX_OK;
+        }
+        
+        // Evade sources of light, fire and dangerous clonks (clonks are only attacked when in groups or very hungry).
+        for(var danger in this.Target->FindObjects(Find_Or(Find_OCF(OCF_CrewMember | OCF_OnFire), Find_Func("IsLightSource")), this.Target->Find_Distance(75), Find_Exclude(this), Sort_Distance()))
+        {
+            if (!PathFree(this.Target->GetX(), this.Target->GetY(), danger->GetX(), danger->GetY()))
+			    continue;
+		    if ((danger->GetOCF() & OCF_CrewMember) && this.Target->WillAttackPrey())
+			    continue;
+		    var x = -Sign(danger->GetX() - this.Target->GetX() + RandomX(-20, 20));
+		    var y = -Sign(danger->GetY() - this.Target->GetY() + RandomX(-20, 20));
+		    this.move_direction = [x, y];
+		    this.Target->DoSonarWave();
+		    // Loose the attack target, because safety is more important.
+		    this.attack_target = nil;
+		    return FX_OK;
+        }
+        // Fly towards attack target when the bat has one.
+	    if (this.attack_target)
```

Comment:
It seems there was a problem with copying the function content into the functions in the proplists.